src/System.Buffers/src/System/Buffers/ArrayPool.cs

```diff
@@ -0,0 +1,97 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Reflection;
+using System.Runtime.InteropServices;
+using System.Threading;
+
+namespace System.Buffers
+{
+    /// <summary>
+    /// This class provides resource pooling for arrays of any type. This is can increase performance
+    /// of certain applications where lots of arrays are created and destroyed in rapid succession.
+    /// This class is thread-safe.
+    /// </summary>
+    public abstract class ArrayPool<T>
+    {
+        /// <summary>The default number of arrays that are available for rent.</summary>
+        public const int DefaultNumberOfArraysPerBucket = 50;
+        /// <summary>The default length of each Rent'able array; equal to 1MB in bytes</summary>
+        public const int DefaultArrayLength = 1024000;
+
+        private static ArrayPool<T> _sharedInstance = null;
+
+        /// <summary>
+        /// Retrieves a shared instance pool that is thread-safe and can be used by 
+        /// multiple components of the same system.
+        /// </summary>
+        public static ArrayPool<T> Shared
+        {
+            get
+            {
+                if (Volatile.Read(ref _sharedInstance) == null)
```

Comment:
Rather than read the field twice in the common case, I'd suggest storing into a local, e.g.
```C#
ArrayPool<T> instance = Volatile.Read(_sharedInstance);
if (instance == null)
{
    Interlocked.CompareExchange(ref _sharedInstance, ...);
    instance = _sharedInstance;
}
return instance;
```